File: solutions/typescript/pascals-triangle/1/pascals-triangle.ts
export class Triangle {
  private triangle: number[][] = [];

  constructor(rows: number) {
    if (rows < 0) {
      throw new Error('Number of rows cannot be negative');
    }
    this.triangle = this.calculateRows(rows);
  }

  private calculateRows(rowCount: number): number[][] {
    if (rowCount === 0) return [];
    
    const result: number[][] = [[1]];
    
    for (let i = 1; i < rowCount; i++) {
      const prevRow = result[i - 1];
      const newRow: number[] = [1];
      
      for (let j = 1; j < i; j++) {
        newRow.push(prevRow[j - 1] + prevRow[j]);
      }
      
      newRow.push(1);
      result.push(newRow);
    }
    
    return result;
  }

  public get lastRow(): number[] {
    return this.triangle.length > 0 ? this.triangle[this.triangle.length - 1] : [];
  }

  public get rows(): number[][] {
    return this.triangle;
  }
}

File: solutions/ruby/pascals-triangle/1/pascals_triangle.rb
class Triangle
  def initialize(row_count)
    @row_count = row_count
  end

  def rows
    return [] if @row_count <= 0

    # Inicializamos el triángulo con la primera fila
    triangle = [[1]]

    # Generamos el resto de las filas de forma consecutiva
    (@row_count - 1).times do
      last_row = triangle.last
      
      # Creamos los dos arreglos desplazados agregando ceros
      left_shift  = last_row + [0]
      right_shift = [0] + last_row

      # Sumamos ambos arreglos elemento por elemento para formar la nueva fila
      next_row = left_shift.zip(right_shift).map { |a, b| a + b }
      
      triangle << next_row
    end

    triangle
  end
end